### PR TITLE
Gideon fix optimizer loading erases parameters

### PR DESCRIPTION
**Pull request by gwenniger**
Dear Sasha and others, I found a bug in the loading of the optimizers from a checkpoint. What happens is that the current code, in the train.py script, completely overwrites the optimizer state even when an existing optimizer is loaded. This happens because the method 
"optim.set_parameters(model.named_parameters())" which calls the optimizer constructor, is always called, even when loading from a checkpoint. This constructor in turn completely resets the state, that is, it completely undoes whatever had been loaded by 
" optim.optimizer.load_state_dict(
            checkpoint['optim'].optimizer.state_dict())" earlier in the 
build_optim(model, checkpoint) method.

I fixed this, by executing the method optim.set_parameters(model.named_parameters())
only when a new optimizer is needed since it is not loaded from a checkpoint. I also added a test to check the state is not empty if an existing optimizer was loaded, which at least for Adam always needs to be true. You can check for yourself, that in the current version of the code this check will not succeed: the state becomes empty when (always) creating a new optimizer in the current version of the code. Note that this bug may have had serious repercussions when running Adam in particular and (often) restarting it from earlier checkpoints.

### CODE REVIEW

**Comment by vince62s** [comment]
If this is the case as it seems, it needs to be fixed, yes.
However, sometimes it is good to reset the states when the model has started to converge and you want to decay and start again. There is a paper on that from Michael Denkowski I think.

On another note, somehow related, as of today, the train_from takes everything from the checkpoint (well except the adam states..) but one may want to change some settings, like the learning rate, the decay method / value, .... which is not overidden by the new opts.
Could be good to allow that.

**Comment by gwenniger** [comment]
Hi Vincent, I agree with your point that resetting the state could be beneficial sometimes, but you don't want this to happen arbitrarily. Re-implementing Denowski's method would be nice, but right now Adam is just not working according to the specification people expect. As such in some of my experiments it sometimes even under-performs compared to SGD, and also converges really slow, things which might both be due to not saving the state when re-loading from saved checkpoints.

**Comment by gwenniger** [comment]
Hi Vincent and others, actually the total fix was a bit more complicated then what I thought earlier.
A similar problem is discussed in the discussion on https://github.com/pytorch/pytorch/issues/2830 .
Essentially it is still necessary to always re-create the optimizer, using the model state, otherwise it will not actually optimize the model parameters and no learning will take place. So the method 
optim.set_parameters(model.parameters()) should be called. 
However, the optimizer parameters from the previous should also be loaded to the optimizer if an optimizer is re-created. To do this, it is necessary to temporarily store these parameters in an additional variable, and then call the method "optim.optimizer.load_state_dict(saved_optimizer_state_dict)" at the right moment, after the optimizer has been re-created. Finally, the thus loaded optimizer state variables should be re-converted to CUDA variables to make this work. All this steps have now been implemented, but there is still something that does not work in some of the  tests appearantly.

**Comment by gwenniger** [comment]
Ok, fixed this now: must use method 
optim.set_parameters(model.**named_parameters**())
not
optim.set_parameters(model.**parameters**())
in method build_optim(model, checkpoint) in train.py

I think everything is working properly now.

**Comment by srush** [comment]
Thanks, I will read this and get it checked in. I agree this is an issue. We need to figure out the semantics of `-train_from` as currently it doesn't allow you to chance optimizer params. 

One other thing. We are using Adam for transformer, and found it necessary to use `-param_init_glorot`. You may want to try that in your experiments.
